packages/react/src/prefabs/PreJoin.tsx

```diff
@@ -59,46 +60,117 @@ export function usePreviewTracks(
+
+  // Update refs when state changes
   React.useEffect(() => {
-    let needsCleanup = false;
-    let localTracks: Array<LocalTrack> = [];
-    trackLock.lock().then(async (unlock) => {
-      try {
-        if (options.audio || options.video) {
-          localTracks = await createLocalTracks(options);
-
-          if (needsCleanup) {
-            localTracks.forEach((tr) => tr.stop());
+    audioTrackRef.current = audioTrack;
+  }, [audioTrack]);
+
+  React.useEffect(() => {
+    videoTrackRef.current = videoTrack;
+  }, [videoTrack]);
+
+  // Shared function to handle track creation and cleanup
+  const handleTrackCreation = React.useCallback(
+    (
+      trackType: 'audio' | 'video',
+      trackOption: any | false,
+      setTrack: React.Dispatch<React.SetStateAction<LocalTrack | undefined>>,
+    ) => {
+      if (!trackOption) {
+        const currentTrack = trackType === 'audio' ? audioTrackRef.current : videoTrackRef.current;
+
+        if (currentTrack) {
+          setOrphanTracks((prev) => [...prev, currentTrack]);
+          setTrack(undefined);
+        }
+        return;
+      }
+
+      let needsCleanup = false;
+      let localTrack: LocalTrack | undefined;
+
+      trackLock.lock().then(async (unlock) => {
+        try {
+          const trackOptions = {
+            audio: trackType === 'audio' ? trackOption : false,
+            video: trackType === 'video' ? trackOption : false,
+          };
+
+          const tracks = await createLocalTracks(trackOptions);
+
+          localTrack = tracks.find((track) => track.kind === trackType);
+
+          if (needsCleanup && localTrack) {
+            localTrack.stop();
+          } else if (localTrack) {
+            // Stop previous track if it exists
+            const currentTrack =
+              trackType === 'audio' ? audioTrackRef.current : videoTrackRef.current;
+
+            if (currentTrack) {
+              setOrphanTracks((prev) => [...prev, currentTrack]);
+            }
+            setTrack(localTrack);
+          }
+        } catch (e: unknown) {
+          if (onError && e instanceof Error) {
+            onError(e);
           } else {
-            setTracks(localTracks);
+            log.error(e);
           }
+        } finally {
+          unlock();
         }
-      } catch (e: unknown) {
-        if (onError && e instanceof Error) {
-          onError(e);
-        } else {
-          log.error(e);
+      });
+
+      return () => {
+        needsCleanup = true;
+        if (localTrack) {
+          localTrack.stop();
         }
-      } finally {
-        unlock();
-      }
-    });
+      };
+    },
+    [trackLock, onError, setOrphanTracks],
+  );
 
-    return () => {
-      needsCleanup = true;
-      localTracks.forEach((track) => {
-        track.stop();
-      });
-    };
-  }, [JSON.stringify(options, roomOptionsStringifyReplacer), onError, trackLock]);
+  // Handle audio track
+  React.useEffect(() => {
```

Comment:
the way it was previously implemented was in order to avoid separate permission prompts when entering the site. 
With your proposed implementation it would result in two different permission requests. 

I don't think that's an improvement worth the additional complexity.